recommendation-app/README.md

```diff
@@ -1,41 +1,338 @@
+) WITH (
+  'connector' = 'kafka', 
+  'topic' = 'flink.sales.records', 
+  'properties.bootstrap.servers' = 'my-cluster-kafka-bootstrap.flink.svc:9092',
+  'properties.group.id' = 'sales-record-group', 
+  'value.format' = 'avro-confluent', 
+  'value.avro-confluent.url' = 'http://apicurio-registry-service.flink.svc:8080/apis/ccompat/v6', 
+  'scan.startup.mode' = 'latest-offset' 
+); 
+```
+
+### Making the recommendation
+
+Now we have the external data accessible via Tables we can put together the recommendation pipeline.
+Lets remind ourselves of what the pipeline needs to recommend:
+
 - Up to 6 highly rated products in the same category as that of the product the customer is currently viewing
 - Only products that are in stock
 - Products that the customer has bought before should be favoured
-- <i>TODO: Avoid showing suggestions that have already been made in previous pageviews </i>
 
-The data:
-- Input: A clickstream (user id, product id, event time)
-- Input: A stream of purchases (user id, product id, purchase date)
-- Input: An inventory of products (product id, product name, product category, number in stock, rating)
+We start with the what the customer is currently viewing, the clickstream.
+But that stream only contains the user ID and the product ID, we need to know the category the product belongs too.
+To do that, we can join the click stream table with the product inventory table to pull out the category.
+
+```sql
+CREATE TEMPORARY VIEW clicked_products AS 
+  SELECT DISTINCT 
+    c.user_id, c.event_time, 
+    p.product_id, 
+    p.category 
+  FROM ClickStreamTable AS c 
+  JOIN ProductInventoryTable AS p ON c.product_id = p.product_id; 
+```
+
+The query above constructs a `TEMPORARY` `VIEW`. 
+Views provide a way to reference the results of a query.
+In the statement above we are assigning the results of the join between the clickstream and product inventory to the `clicked_products` view, so we can reference it in later queries.
+The `TEMPORARY` clause in the `CREATE` statement refers to how Flink will persist the metadata about this view and is linked to Flink Catalogs that we won't cover here (you can read more about Catalogs in the [docs](https://nightlies.apache.org/flink/flink-docs-stable/docs/dev/table/catalogs/)).
+
+Now we have the product category information in the clickstream we need to get all products, in that category, that are in stock and mark weather the user purchased them or not.
+To do that we can `JOIN` the enriched clickstream on the product inventory table, where the categories match, to get all products in that category. 
+This allows us to also pull in the stock level and ratings. 
+We can then take this set of products from the same category and `LEFT JOIN` on the sales records table. 
+This will pass through all entries from the category, weather they have been purchased or not and allow us to have a new `purchased` column for any products the the user has brought in the past.  
```

Comment:
```suggestion
This will pass through all entries from the category, whether they have been purchased or not and allow us to have a new `purchased` column for any products the the user has bought in the past.  
```